langchain/src/vectorstores/prisma.ts

```diff
@@ -0,0 +1,249 @@
+    )?.[0];
+
+    if (idColumn == null) throw new Error("Missing ID column");
+    if (contentColumn == null) throw new Error("Missing content column");
+
+    this.idColumn = idColumn;
+    this.contentColumn = contentColumn;
+
+    this.tableSql = this.Prisma.raw(`"${config.tableName}"`);
+    this.vectorColumnSql = this.Prisma.raw(`"${config.vectorColumnName}"`);
+
+    this.selectSql = this.Prisma.raw(
+      entries
+        .map(([key, alias]) => (alias && key) || null)
+        .filter((x): x is string => !!x)
+        .map((key) => `"${key}"`)
+        .join(", ")
+    );
+  }
+
+  static withModel<TModel extends Record<string, unknown>>(db: PrismaClient) {
+    function create<
+      TPrisma extends PrismaNamespace,
+      TColumns extends ModelColumns<TModel>
+    >(
+      config: {
+        prisma: TPrisma;
+        tableName: keyof TPrisma["ModelName"] & string;
+        vectorColumnName: string;
+        columns: TColumns;
+      },
+      embeddings: Embeddings
+    ) {
+      type ModelName = keyof TPrisma["ModelName"] & string;
+      return new PrismaVectorStore<TModel, ModelName, TColumns>(
+        { db, ...config },
+        embeddings
+      );
+    }
+
+    return { create };
+  }
+
+  async addModels(models: TModel[]) {
+    return this.addDocuments(
+      models.map((metadata) => {
+        const pageContent = typeof metadata[this.contentColumn];
+        if (pageContent !== "string")
+          throw new Error("Content column must be a string");
+        return new Document({ pageContent, metadata });
+      })
+    );
+  }
+
+  async addDocuments(documents: Document<TModel>[]) {
+    const texts = documents.map(({ pageContent }) => pageContent);
+    return this.addVectors(
+      await this.embeddings.embedDocuments(texts),
+      documents
+    );
+  }
+
+  async addVectors(vectors: number[][], documents: Document<TModel>[]) {
+    const idSql = this.Prisma.raw(`"${this.idColumn}"`);
+
+    await this.db.$transaction(
+      vectors.map(
+        (vector, idx) => this.db.$executeRaw`
+          UPDATE ${this.tableSql}
+          SET ${this.vectorColumnSql} = ${`[${vector.join(",")}]`}::vector
+          WHERE ${idSql} = ${documents[idx].metadata[this.idColumn]}
+        `
+      )
+    );
+  }
+
+  async similaritySearch(
+    query: string,
+    k = 4
+  ): Promise<Document<SimilarityModel<TModel, TSelectModel>>[]> {
+    const results = await this.similaritySearchVectorWithScore(
+      await this.embeddings.embedQuery(query),
+      k
+    );
+
+    return results.map((result) => result[0]);
+  }
+
+  async similaritySearchVectorWithScore(
+    query: number[],
+    k: number
+  ): Promise<[Document<SimilarityModel<TModel, TSelectModel>>, number][]> {
+    const vectorQuery = `[${query.join(",")}]`;
+    const articles = await this.db.$queryRaw<
+      Array<SimilarityModel<TModel, TSelectModel>>
+    >`
+      SELECT ${this.selectSql}, ${this.vectorColumnSql} <=> ${vectorQuery}::vector as "_distance" 
```

Comment:
will prisma escape these interpolated values? otherwise could be potentially dangerous

Reply:
I believe that should be the case for `vectorQuery`, as the query has been built using `$queryRaw` instead of `$queryRawUnsafe` and will use parametrised queries: https://github.com/prisma/prisma-client-js/issues/727#issuecomment-650096790

However, `this.selectSql` (coming from `dbConfig.columns`), `this.vectorColumnSql` (coming from `dbConfig.vectorColumnName`) and `this.tableSql` (coming from `dbConfig.tableName`). 

Assuming these values come strictly from the developer and _not_ from user input, I believe it would be acceptable to leave it as-is and add a disclaimer to shift the responsibility of sanitising to the consumer. What do you think @nfcampos ?

Reply:
Okay great!